snuba/datasets/discover.py

```diff
@@ -0,0 +1,278 @@
+from snuba.datasets.factory import get_dataset
+from snuba.datasets.schemas import Schema, RelationalSource
+from snuba.datasets.schemas.tables import TableSource
+from snuba.query.extensions import QueryExtension
+from snuba.query.parsing import ParsingContext
+from snuba.query.project_extension import ProjectExtension, ProjectWithGroupsProcessor
+from snuba.query.query import Query
+from snuba.query.query_processor import QueryProcessor
+from snuba.query.timeseries import TimeSeriesExtension
+from snuba.query.types import Condition
+from snuba.request.request_settings import RequestSettings
+from snuba.util import is_condition
+
+EVENTS = 'events'
+TRANSACTIONS = 'transactions'
+
+
+def detect_dataset(query: Query, transactions_columns: ColumnSet) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
+
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    if [col for col in query.get_columns_referenced_in_conditions() if transactions_columns.get(col)]:
+        return TRANSACTIONS
+
+    # Use events by default
+    return EVENTS
+
+
+class DiscoverTableSource:
+    def __init__(self):
+        self.__table_source = None
+
+    def get_table_source(self) -> Optional[TableSource]:
+        return self.__table_source
+
+    def set_table_source(self, dataset_name: str) -> None:
+        self.__table_source = get_dataset(dataset_name) \
+            .get_dataset_schemas() \
+            .get_read_schema() \
+            .get_data_source()
+
+
+class DiscoverSource(RelationalSource):
+    def __init__(self, columns: ColumnSet, discover_table_source: DiscoverTableSource) -> None:
+        self.__columns = columns
+        self.__discover_table_source = discover_table_source
+
+    def format_from(self) -> str:
+        table_source = self.__discover_table_source.get_table_source()
+        if table_source:
+            return table_source.format_from()
+        return ""
+
+    def get_columns(self) -> ColumnSet:
+        return self.__columns
+
+    def get_mandatory_conditions(self) -> Sequence[Condition]:
+        table_source = self.__discover_table_source.get_table_source()
+        if table_source:
+            return table_source.get_mandatory_conditions()
+        return []
+
+
+class DiscoverProcessor(QueryProcessor):
```

Comment:
Processors should be relatively simple rules applied in sequence to a query. Could you please give a more specific name to this one, since likely there won't be a single big processor for Discover.